File: src/storage.rs
use anyhow::Result;
use chrono::{DateTime, Utc};
use sqlx::{postgres::PgPoolOptions, PgPool, Row};
use uuid::Uuid;

use crate::summarize::SummaryResult;
use crate::transcribe::TranscriptionResult;

#[derive(Clone)]
pub struct Database {
    pool: PgPool,
}

#[derive(Clone, Debug)]
pub struct SummaryRecord {
    pub title: Option<String>,
    pub source_url: String,
    pub published_at: Option<DateTime<Utc>>,
    pub summary: String,
    pub key_points: Vec<String>,
    pub created_at: DateTime<Utc>,
}

impl Database {
    pub async fn connect(database_url: &str) -> Result<Self> {
        let pool = PgPoolOptions::new()
            .max_connections(10)
            .connect(database_url)
            .await?;

        Ok(Self { pool })
    }

    pub async fn migrate(&self) -> Result<()> {
        sqlx::query(
            "\
            CREATE TABLE IF NOT EXISTS feeds (
                id UUID PRIMARY KEY,
                url TEXT NOT NULL UNIQUE,
                title TEXT,
                last_checked TIMESTAMPTZ
            );
            ",
        )
        .execute(&self.pool)
        .await?;

        sqlx::query(
            "\
            CREATE TABLE IF NOT EXISTS videos (
                id UUID PRIMARY KEY,
                feed_id UUID REFERENCES feeds(id),
                source_url TEXT NOT NULL UNIQUE,
                guid TEXT,
                title TEXT,
                published_at TIMESTAMPTZ
            );
            ",
        )
        .execute(&self.pool)
        .await?;

        sqlx::query(
            "\
            CREATE TABLE IF NOT EXISTS transcripts (
                id UUID PRIMARY KEY,
                video_id UUID NOT NULL REFERENCES videos(id),
                language TEXT,
                confidence REAL,
                text TEXT NOT NULL,
                created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
            );
            ",
        )
        .execute(&self.pool)
        .await?;

        sqlx::query(
            "\
            CREATE TABLE IF NOT EXISTS summaries (
                id UUID PRIMARY KEY,
                video_id UUID NOT NULL REFERENCES videos(id),
                summary TEXT NOT NULL,
                key_points JSONB NOT NULL,
                created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
            );
            ",
        )
        .execute(&self.pool)
        .await?;

        Ok(())
    }

    pub async fn upsert_feed(&self, url: &str, title: Option<&str>) -> Result<Uuid> {
        let id = Uuid::new_v4();
        let row = sqlx::query(
            "\
            INSERT INTO feeds (id, url, title, last_checked)
            VALUES ($1, $2, $3, NOW())
            ON CONFLICT (url)
            DO UPDATE SET title = COALESCE(EXCLUDED.title, feeds.title), last_checked = NOW()
            RETURNING id;
            ",
        )
        .bind(id)
        .bind(url)
        .bind(title)
        .fetch_one(&self.pool)
        .await?;

        Ok(row.try_get("id")?)
    }

    pub async fn upsert_video(
        &self,
        feed_id: Option<Uuid>,
        guid: Option<&str>,
        title: Option<&str>,
        source_url: &str,
        published_at: Option<DateTime<Utc>>,
    ) -> Result<Uuid> {
        let id = Uuid::new_v4();
        let row = sqlx::query(
            "\
            INSERT INTO videos (id, feed_id, source_url, guid, title, published_at)
            VALUES ($1, $2, $3, $4, $5, $6)
            ON CONFLICT (source_url)
            DO UPDATE SET
                title = COALESCE(EXCLUDED.title, videos.title),
                published_at = COALESCE(EXCLUDED.published_at, videos.published_at),
                guid = COALESCE(EXCLUDED.guid, videos.guid),
                feed_id = COALESCE(EXCLUDED.feed_id, videos.feed_id)
            RETURNING id;
            ",
        )
        .bind(id)
        .bind(feed_id)
        .bind(source_url)
        .bind(guid)
        .bind(title)
        .bind(published_at)
        .fetch_one(&self.pool)
        .await?;

        Ok(row.try_get("id")?)
    }

    pub async fn insert_transcript(
        &self,
        video_id: Uuid,
        transcript: &TranscriptionResult,
    ) -> Result<Uuid> {
        let id = Uuid::new_v4();
        sqlx::query(
            "\
            INSERT INTO transcripts (id, video_id, language, confidence, text)
            VALUES ($1, $2, $3, $4, $5);
            ",
        )
        .bind(id)
        .bind(video_id)
        .bind(&transcript.language)
        .bind(transcript.confidence)
        .bind(&transcript.text)
        .execute(&self.pool)
        .await?;

        Ok(id)
    }

    pub async fn insert_summary(&self, video_id: Uuid, summary: &SummaryResult) -> Result<Uuid> {
        let id = Uuid::new_v4();
        let key_points = serde_json::to_value(&summary.key_points)?;
        sqlx::query(
            "\
            INSERT INTO summaries (id, video_id, summary, key_points)
            VALUES ($1, $2, $3, $4);
            ",
        )
        .bind(id)
        .bind(video_id)
        .bind(&summary.summary)
        .bind(key_points)
        .execute(&self.pool)
        .await?;

        Ok(id)
    }

    pub async fn latest_summaries(&self, limit: i64) -> Result<Vec<SummaryRecord>> {
        let rows = sqlx::query(
            "\
            SELECT v.title, v.source_url, v.published_at, s.summary, s.key_points, s.created_at
            FROM summaries s
            JOIN videos v ON v.id = s.video_id
            ORDER BY s.created_at DESC
            LIMIT $1;
            ",
        )
        .bind(limit)
        .fetch_all(&self.pool)
        .await?;

        let mut items = Vec::with_capacity(rows.len());
        for row in rows {
            let key_points_value: serde_json::Value = row.try_get("key_points")?;
            let key_points = key_points_value
                .as_array()
                .map(|arr| {
                    arr.iter()
                        .filter_map(|item| item.as_str().map(|value| value.to_string()))
                        .collect::<Vec<_>>()
                })
                .unwrap_or_default();

            items.push(SummaryRecord {
                title: row.try_get("title")?,
                source_url: row.try_get("source_url")?,
                published_at: row.try_get("published_at")?,
                summary: row.try_get("summary")?,
                key_points,
                created_at: row.try_get("created_at")?,
            });
        }

        Ok(items)
    }
}

File: vra/storage.py
from __future__ import annotations

import json
import logging
import uuid
from dataclasses import dataclass
from datetime import datetime

import asyncpg

from .transcribe import TranscriptionResult
from .summarize import SummaryResult

log = logging.getLogger(__name__)

_SCHEMA = """\
CREATE TABLE IF NOT EXISTS feeds (
    id UUID PRIMARY KEY,
    url TEXT NOT NULL UNIQUE,
    title TEXT,
    last_checked TIMESTAMPTZ
);
CREATE TABLE IF NOT EXISTS videos (
    id UUID PRIMARY KEY,
    feed_id UUID REFERENCES feeds(id),
    source_url TEXT NOT NULL UNIQUE,
    guid TEXT,
    title TEXT,
    published_at TIMESTAMPTZ
);
CREATE TABLE IF NOT EXISTS transcripts (
    id UUID PRIMARY KEY,
    video_id UUID NOT NULL REFERENCES videos(id),
    language TEXT,
    text TEXT NOT NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
);
CREATE TABLE IF NOT EXISTS summaries (
    id UUID PRIMARY KEY,
    video_id UUID NOT NULL REFERENCES videos(id),
    summary TEXT NOT NULL,
    key_points JSONB NOT NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
);"""


@dataclass(slots=True)
class SummaryRecord:
    title: str | None
    source_url: str
    published_at: datetime | None
    summary: str
    key_points: list[str]
    created_at: datetime


class Database:
    def __init__(self, pool: asyncpg.Pool) -> None:
        self._pool = pool

    @classmethod
    async def connect(cls, database_url: str) -> Database:
        pool = await asyncpg.create_pool(database_url, min_size=2, max_size=10)
        return cls(pool)

    async def migrate(self) -> None:
        async with self._pool.acquire() as conn:
            await conn.execute(_SCHEMA)
        log.info("Database schema ready")

    async def upsert_feed(self, url: str, title: str | None) -> uuid.UUID:
        fid = uuid.uuid4()
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                """INSERT INTO feeds (id, url, title, last_checked)
                   VALUES ($1, $2, $3, NOW())
                   ON CONFLICT (url)
                   DO UPDATE SET title = COALESCE(EXCLUDED.title, feeds.title),
                                 last_checked = NOW()
                   RETURNING id""",
                fid, url, title,
            )
        return row["id"]

    async def upsert_video(
        self,
        feed_id: uuid.UUID | None,
        guid: str | None,
        title: str | None,
        source_url: str,
        published_at: datetime | None,
    ) -> uuid.UUID:
        vid = uuid.uuid4()
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(
                """INSERT INTO videos (id, feed_id, source_url, guid, title, published_at)
                   VALUES ($1, $2, $3, $4, $5, $6)
                   ON CONFLICT (source_url)
                   DO UPDATE SET
                       title = COALESCE(EXCLUDED.title, videos.title),
                       published_at = COALESCE(EXCLUDED.published_at, videos.published_at),
                       guid = COALESCE(EXCLUDED.guid, videos.guid),
                       feed_id = COALESCE(EXCLUDED.feed_id, videos.feed_id)
                   RETURNING id""",
                vid, feed_id, source_url, guid, title, published_at,
            )
        return row["id"]

    async def insert_transcript(self, video_id: uuid.UUID, tr: TranscriptionResult) -> uuid.UUID:
        tid = uuid.uuid4()
        async with self._pool.acquire() as conn:
            await conn.execute(
                """INSERT INTO transcripts (id, video_id, language, text)
                   VALUES ($1, $2, $3, $4)""",
                tid, video_id, tr.language, tr.text,
            )
        return tid

    async def insert_summary(self, video_id: uuid.UUID, sr: SummaryResult) -> uuid.UUID:
        sid = uuid.uuid4()
        async with self._pool.acquire() as conn:
            await conn.execute(
                """INSERT INTO summaries (id, video_id, summary, key_points)
                   VALUES ($1, $2, $3, $4::jsonb)""",
                sid, video_id, sr.summary, json.dumps(sr.key_points),
            )
        return sid

    async def latest_summaries(self, limit: int = 20) -> list[SummaryRecord]:
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(
                """SELECT v.title, v.source_url, v.published_at,
                          s.summary, s.key_points, s.created_at
                   FROM summaries s
                   JOIN videos v ON v.id = s.video_id
                   ORDER BY s.created_at DESC
                   LIMIT $1""",
                limit,
            )
        out: list[SummaryRecord] = []
        for r in rows:
            kp = r["key_points"]
            if isinstance(kp, str):
                kp = json.loads(kp)
            out.append(SummaryRecord(
                title=r["title"],
                source_url=r["source_url"],
                published_at=r["published_at"],
                summary=r["summary"],
                key_points=kp if isinstance(kp, list) else [],
                created_at=r["created_at"],
            ))
        return out
